### PR TITLE
Baai glm

### PR DESCRIPTION
**Pull request by shunxing1234**
add glm model

### CODE REVIEW

File: src/transformers/models/glm/configuration_glm.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+# limitations under the License.
+""" GLM model configuration """
+
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+logger = logging.get_logger(__name__)
+
+GLM_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "shunxing1234/GLM-base-cased": "https://huggingface.co/shunxing1234/GLM-base-cased/resolve/main/config.json",
```
Comment (marscrazy) [inline src/transformers/models/glm/configuration_glm.py:23]: GLM large？

File: src/transformers/models/glm/__init__.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    if is_tokenizers_available():
+        from .tokenization_glm_fast import GLMTokenizerFast
+
+    if is_torch_available():
+        from .modeling_glm import (
+            GLM_PRETRAINED_MODEL_ARCHIVE_LIST,
+            GLMForMaskedLM,
+            GLMForCausalLM,
+            GLMForMultipleChoice,
+            GLMForQuestionAnswering,
```
Comment (marscrazy) [inline src/transformers/models/glm/__init__.py:64]: remove unused classes

File: src/transformers/models/glm/__init__.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+if is_tokenizers_available():
+    _import_structure["tokenization_glm_fast"] = ["GLMTokenizerFast"]
+
+if is_torch_available():
+    _import_structure["modeling_glm"] = [
+        "GLM_PRETRAINED_MODEL_ARCHIVE_LIST",
+        "GLMForMaskedLM",
+        "GLMForCausalLM",
```
Comment (marscrazy) [inline src/transformers/models/glm/__init__.py:37]: clean

File: src/transformers/models/glm/configuration_glm.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            output_dropout_prob=0.1,
+            max_sequence_length=512,
+            max_memory_length=0,
+            checkpoint_activations=False,
+            checkpoint_num_layers=1,
+            parallel_output=True,
+            relative_encoding=False,
+            block_position_encoding=True,
+            output_predict=False,
+            spell_length=None,
```
Comment (marscrazy) [inline src/transformers/models/glm/configuration_glm.py:110]: p-tuning ?

File: src/transformers/models/glm/modeling_glm.py [hunk 5]
```diff
@@ -0,0 +1,998 @@
...
+from ...activations import ACT2FN
+from ...utils import (
+    add_code_sample_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    replace_return_docstrings,
+)
+from ...modeling_outputs import (
+    BaseModelOutputWithPastAndCrossAttentions,
+    CausalLMOutputWithCrossAttentions,
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:37]: clean

File: src/transformers/models/glm/modeling_glm.py [hunk 6]
```diff
@@ -0,0 +1,998 @@
...
+    # Split.
+    tensor_list = torch.split(tensor, last_dim_size, dim=last_dim)
+    # Note: torch.split does not create contiguous tensors by default.
+    if contiguous_split_chunks:
+        return tuple(chunk.contiguous() for chunk in tensor_list)
+
+    return tensor_list
+
+
+def gelu_impl(x):
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:120]: transformers/activations.py

File: src/transformers/models/glm/modeling_glm.py [hunk 7]
```diff
@@ -0,0 +1,998 @@
...
+        self.sparse = False
+        self._weight = None
+
+        self.vocab_start_index = 0
+        self.vocab_end_index = self.num_embeddings
+
+        # Allocate weights.
+        self.weight = Parameter(torch.Tensor(self.num_embeddings,
+                                             self.embedding_dim))
+        self.weight.model_parallel = True
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:206]: remove

File: src/transformers/models/glm/modeling_glm.py [hunk 8]
```diff
@@ -0,0 +1,998 @@
...
+        b: batch size
+        s: sequence length
+    """
+
+    def __init__(self, hidden_size, num_attention_heads,
+                 attention_dropout_prob, output_dropout_prob,
+                 init_method, output_layer_init_method=None, relative_encoding=False,
+                 performer=False, attention_scale=1.0):
+        super(SelfAttention, self).__init__()
+        self.performer = performer
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:271]: performer？

File: src/transformers/models/glm/modeling_glm.py [hunk 9]
```diff
@@ -0,0 +1,998 @@
...
+        return x
+
+    def forward(self, hidden_states, ltor_mask, position_embeddings=None, r_w_bias=None, r_r_bias=None, mem=None):
+        # hidden_states: [b, s, h]
+        # ltor_mask: [1, 1, s, s]
+
+        # Attention heads. [b, s, hp]
+        query_length = hidden_states.size(1)
+
+        if mem is None:
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:333]: add instruction

File: src/transformers/models/glm/modeling_glm.py [hunk 10]
```diff
@@ -0,0 +1,998 @@
...
+
+        if zero_triu:
+            ones = torch.ones((x.size(0), x.size(1)))
+            x = x * torch.tril(ones, x.size(1) - x.size(0))[:, :, None, None]
+
+        return x
+
+    def forward(self, hidden_states, ltor_mask, position_embeddings=None, r_w_bias=None, r_r_bias=None, mem=None):
+        # hidden_states: [b, s, h]
+        # ltor_mask: [1, 1, s, s]
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:328]: [b,1, s,s]

File: src/transformers/models/glm/modeling_glm.py [hunk 11]
```diff
@@ -0,0 +1,998 @@
...
+        # MLP
+        self.mlp = MLP(
+            hidden_size,
+            output_dropout_prob,
+            init_method,
+            output_layer_init_method=output_layer_init_method)
+
+    def forward(self, hidden_states, ltor_mask, position_embeddings=None, r_w_bias=None, r_r_bias=None, mem=None):
+        # hidden_states: [b, s, h]
+        # ltor_mask: [1, 1, s, s]
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:477]: [b,1,s,s]

File: src/transformers/models/glm/modeling_glm.py [hunk 12]
```diff
@@ -0,0 +1,998 @@
...
+        layernorm_output = self.post_attention_layernorm(layernorm_input)
+        # MLP.
+        mlp_output = self.mlp(layernorm_output)
+        # Second residual connection.
+        output = layernorm_input + mlp_output
+
+        return output
+
+
+class Transformer(torch.nn.Module):
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:496]: stack

File: src/transformers/models/glm/modeling_glm.py [hunk 13]
```diff
@@ -0,0 +1,998 @@
...
+        context_layer = context_layer.view(*new_context_layer_shape)
+
+        # Output. [b, s, h]
+        output = self.dense(context_layer)
+        output = self.output_dropout(output)
+
+        return output
+
+
+class TransformerLayer(torch.nn.Module):
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:404]: block

File: src/transformers/models/glm/modeling_glm.py [hunk 14]
```diff
@@ -0,0 +1,998 @@
...
+
+        # Transformer layers.
+        self.layers = torch.nn.ModuleList(
+            [get_layer() for _ in range(num_layers)])
+
+        # Final layer norm before output.
+        self.final_layernorm = LayerNorm(hidden_size, eps=layernorm_epsilon)
+
+    def forward(self, hidden_states, position_ids, attention_mask, memory_states=None, encoder_states=None,
+                return_memory=False, detach_memory=True):
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:619]: describe following parameters:

memory_states=None, encoder_states=None,
                 return_memory=False, detach_memory=True

File: src/transformers/models/glm/modeling_glm.py [hunk 15]
```diff
@@ -0,0 +1,998 @@
...
+                if memory_length > 0:
+                    m = m.expand(batch_size, -1, -1)
+                    m = torch.cat((hidden_states.new_ones((batch_size, seq_length, memory_length)), m), dim=2)
+                m = m.unsqueeze(1)
+                return m
+
+            if not self.performer:
+                attention_mask = build_mask_matrix(query_length, sep, memory_length=memory_length)
+        else:
+            attention_mask = attention_mask[:, :, :, -query_length - memory_length:]
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:655]: add a shape check for attention_mask [b,1,s,s]

File: src/transformers/models/glm/modeling_glm.py [hunk 16]
```diff
@@ -0,0 +1,998 @@
...
+                                       config.checkpoint_activations,
+                                       config.checkpoint_num_layers,
+                                       attention_scale=config.attention_scale,
+                                       relative_encoding=config.relative_encoding,
+                                       block_position_encoding=config.block_position_encoding)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def freeze_transformer(self, tune_prefix_layers=None):
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:870]: remove?

File: src/transformers/models/glm/modeling_glm.py [hunk 17]
```diff
@@ -0,0 +1,998 @@
...
+        checkpoint=_CHECKPOINT_FOR_DOC,
+        output_type=BaseModelOutputWithPastAndCrossAttentions,
+        config_class=_CONFIG_FOR_DOC,
+    )
+    def forward(
+            self,
+            input_ids=None,
+            position_ids=None,
+            attention_mask=None,
+            mems=None,
```
Comment (marscrazy) [inline src/transformers/models/glm/modeling_glm.py:891]: add parameter descriptions

File: src/transformers/models/glm/tokenization_glm.py [hunk 18]
```diff
@@ -0,0 +1,644 @@
...
+        if self.do_basic_tokenize:
+            split_tokens = []
+            for token in self.basic_tokenizer.tokenize(text):
+                for sub_token in self.wordpiece_tokenizer.tokenize(token):
+                    split_tokens.append(sub_token)
+        else:
+            split_tokens = self.wordpiece_tokenizer.tokenize(text)
+        return split_tokens
+
+    def build_input_from_ids(
```
Comment (marscrazy) [inline src/transformers/models/glm/tokenization_glm.py:284]: comments

File: tests/glm/test_modeling_glm.py [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+        self.num_labels = num_labels
+        self.num_choices = num_choices
+        self.scope = scope
+
+    def prepare_config_and_inputs(self):
+        input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
+
+        input_mask = None
+        if self.use_input_mask:
+            input_mask = random_attention_mask([self.batch_size, self.seq_length])
```
Comment (marscrazy) [inline tests/glm/test_modeling_glm.py:96]: check the shape of input_mask 
need to be a scalar or [b,1,s,s] size matrix

File: tests/glm/test_modeling_glm.py [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+        model.to(torch_device)
+
+        torch.manual_seed(0)
+        tokenized = tokenizer("Today is a nice day and", return_tensors="pt", return_token_type_ids=True)
+        input_ids = tokenized.input_ids.to(torch_device)
+        output_ids = model.generate(input_ids, do_sample=True)
+        output_str = tokenizer.decode(output_ids[0], skip_special_tokens=True)
+
+        token_type_ids = tokenized.token_type_ids.to(torch_device)
+        output_seq = model.generate(input_ids=input_ids, do_sample=True, num_return_sequences=5)
```
Comment (marscrazy) [inline tests/glm/test_modeling_glm.py:246]: attention mask : [0]* batch_size

File: tests/utils/test_utils_check_copies.py [hunk 21]
```diff
@@ -123,7 +123,7 @@ def test_is_copy_consistent(self):
         )
 
     def test_convert_to_localized_md(self):
-        localized_readme = check_copies.LOCALIZED_READMES["README_zh-hans.md"]
```
Comment (marscrazy) [inline tests/utils/test_utils_check_copies.py:126]: check
